### DISCUSSION TITLE
Sample Class File?

### DISCUSSION THREAD

**Discussion by bmedwards** [Kentico/xperience-algolia#31] [Q&A]
Please forgive me, as I am primarily a php developer, getting familiar with Kentico 13 and .NET Core together.   

I am attempting to get the Algolia integartion going, but am stuck on creating the custom class file referenced in the project.  I created an included project inside of the Kentico Administation Project (CMSApp) called AlgoliaSearcg, with a class file (AlgoliaSearch.cs) containing the code extending the AlgoliaSearchModel class, however, I am still receiving the error: "Unable to resolve service for type 'Kentico.Xperience.AlgoliaSearch.Services.IAlgoliaIndexRegister' while attempting to activate 'Kentico.Xperience.AlgoliaSearch.Services.DefaultAlgoliaRegistrationService'."

Do you have a sample classfile or project to include in trhe Administration project that I can use to test with the Dancing Goat website?

Appreciate any help with this.  Thanks!

**Comment by seangwright** [comment]
@bmedwards Hey Brian!

You'll want to have a custom `Module` class in a .NET project that your `CMSApp` application references, or in the `CMSApp` application itself. That `Module` class will look something like this:

```csharp
using CMS;
using CMS.Core;
using CMS.DataEngine;
using CMSApp.Configuration;
using Kentico.Xperience.AlgoliaSearch.Services;

[assembly: RegisterModule(typeof(DependencyManagementModule))]

namespace CMSApp.Configuration
{
    public class DependencyManagementModule : Module
    {
        public DependencyManagementModule() : base(nameof(DependencyManagementModule)) { }

        protected override void OnPreInit()
        {
            base.OnPreInit();

            // See: https://github.com/Kentico/xperience-algolia#gear-creating-and-registering-an-algolia-index
            string indexName = "..."; // populate this however you want

			// AlgoliaCustomSearchModel is your custom class that inherits from AlgoliaSearchModel
			IAlgoliaIndexRegister register = new DefaultAlgoliaIndexRegister().Add<AlgoliaCustomSearchModel>(indexName);

            Service.Use<IAlgoliaIndexRegister>(register);
        }

        protected override void OnInit()
        {
            base.OnInit();

            //
        }
    }
}
```

This `DependencyManagementModule` class can be named anything. 

I typically use it as the single location where all custom dependencies and modules are registered that I want the CMS application to use.

---

On the ASPNET Core application side, you'll have some customization to the `IServiceCollection`. I usually put this in an extension method and call it from `Startup.cs`:

```csharp
public static class ServiceCollectionXperienceExtensions
{
    public static IServiceCollection AddXperienceAlgolia(this IServiceCollection services, IConfiguration config)
    {
        var algoliaSection = config.GetSection(AlgoliaOptions.SECTION_NAME);
        string indexName = "..."; // get this however you want

        // This is the same thing we did in the DependencyManagementModule in the CMSApp project
        var register = new DefaultAlgoliaIndexRegister().Add<AlgoliaSiteSearchModel>(indexName);

        return services
            .AddAlgolia(config, register)
            .Configure<AlgoliaOptions>(options =>
            {
                var algoliaOptions = algoliaSection.Get<AlgoliaOptions>();

                options.ApiKey = algoliaOptions.ApiKey;
                options.ApplicationId = algoliaOptions.ApplicationId;
                options.SearchKey = algoliaOptions.SearchKey;
            });
    }
}
```

**Reply by bmedwards** [reply]
Thank you @seangwright for the quick reply.  

I am attempting to add the Module class to the 'CMSApp' .NetCore Solution as a new .Net2.0 Class project, but still getting the same results.  Is that correct?  I added the Kentico.Xperience.Algolia.KX13 and Kentico.Xperience.Libraries to the .Net 2.0 Class as well.

If I were to add it to the CMSApp project itself, where would it go?

Appreciate any help you have.

**Reply by seangwright** [reply]
Your `AlgoliaCustomSearchModel` class is something you'd want in a shared class library which would target `netstandard2.0`.

This library project would be referenced as a dependency by both the `CMSApp` application and the ASPNET Core 6.0 application.

That shared library should have a dependency on `Kentico.Xperience.Algolia.KX13` which includes [Kentico.Xperience.Libraries as a dependency](https://www.nuget.org/packages/Kentico.Xperience.Algolia.KX13#dependencies-body-tab), but you can include `Kentico.Xperience.Libraries` as an explicit package dependency as well.

To reference your shared class library in your `CMSApp` project, follow [the instructions for adding project references in Visual Studio](https://learn.microsoft.com/en-us/visualstudio/ide/managing-references-in-a-project?view=vs-2022). This adds an entry in the `CMSApp.csproj` XML file.

You can do the same thing with your ASP.NET Core 6.0 project in Visual Studio. You could also use the [dotnet add reference](https://learn.microsoft.com/en-us/dotnet/core/tools/dotnet-add-reference) command at the command line using the .NET SDK (which is included with Visual Studio).

You will want to make sure your shared class library has an `[assembly: AssemblyDiscoverable]` attribute somewhere in it. I prefer the approach of adding the `.csproj` entry [as detailed in the Xperience docs](https://docs.xperience.io/custom-development/adding-custom-assemblies).

```xml
<ItemGroup>
    <AssemblyAttribute Include="CMS.AssemblyDiscoverableAttribute">
    </AssemblyAttribute>
</ItemGroup>
```

This ensures that Xperience's code can find all the custom classes and types defined in your shared class library.
